src/protocols/PhysicalLayer.cpp

```diff
@@ -50,29 +50,65 @@ int16_t PhysicalLayer::startTransmit(const char* str, uint8_t addr) {
 
 int16_t PhysicalLayer::readData(String& str, size_t len) {
   // create temporary array to store received data
-  char* data = new char[len + 1];
-  int16_t state = readData((uint8_t*)data, len);
-
-  // if packet was received successfully, copy data into String
-  if(state == ERR_NONE) {
-    str = String(data);
-  }
-
-  delete[] data;
+  uint8_t* data = nullptr;
+  int16_t state = ERR_NONE;
+  
+  if(len == 0)
+  {
+    //We can query the packet length now because the packet should have been received earlier
+    getPacketLength(len);
+  }  
+  
+    //Build a temporary buffer
+    data = new uint8_t[len + 1];
+    // read the received data
+    if(data)
+    {
+      state = (readData(data, len));
+    }
+    else
+    {
+      state = ERR_INVALID_MEMORYBUFFER;
+    }
+    //guards against dynamic allocation issue #51 and #30
```

Comment:
Do not use issue numbers directly in code comments. This is because most code from LoRaLib is directly reused in RadioLib (i.e. copied and pasted; Arduino libraries can't use dependencies), so issue numbers would lead to confusion.

Reply:
Understood!